### PR TITLE
[WIP][Feature](bangc-ops):add new op dynamic_point_to_voxel_forward

### PR DESCRIPTION
**Pull request by tudejiang79**
Thanks for your contribution and we appreciate it a lot. 

## 1. Motivation

Please describe your motivation and the goal you want to achieve through this pull request.

## 2. Modification

Please briefly describe what modification is made in this pull request, and indicate where to make the modification.

## 3. Test Report

If you want to know how to do operator testing, you can see [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).

#### 3.1.1 Accuracy Acceptance Standard

For static threshold standard details, see: [MLU-OPS Accuracy Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Accuracy-Acceptance-Standard.md).

- [ ] diff1: diff1 <= 3e-3
- [ ] diff2: diff2 <= 3e-3

#### 3.1.2 Operator Scheme checklist

|     No.        |                 Details              |            Check Results             |
|----------------|--------------------------------------|--------------------------------------|
|        1       |Supported hardware                    |             MLU370<br>MLU590         |
|        2       |Job types                             |          block <br> U1 <br> U4       |
|        3       |Layouts                               |          NHWC 、NCHW、ARRAY etc      |
|        4       |Whether multi-dimensions are supported|                                      |
|        5       |Whether element zero is supported     |                                      |
|        6       |Data type(half/float)                 |           half / float etc           |
|        7       |Whether there is size limit           |                                      |

#### 3.1.3 New Feature Test

If you have checked the following items, please tick the relevant box.

- [ ] Data type test
- [ ] Multi-dimensional tensor test
- [ ] Layout test
- [ ] Different size/integer remainder end segment/alignment misalignment test
- [ ] Zero dimensional tensor test/zero element test
- [ ] stability test
- [ ] Multiple platform test
- [ ] Gen_case module test
- [ ] Nan/INF tests 
- [ ] Bug fix tests
- [ ] For memory leak check details, see[GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For code coverage check details, see: [GTest-User-Guide-zh](https://github.com/Cambricon/mlu-ops/blob/master/docs/GTest-User-Guide-zh.md).
- [ ] For I/O calculation efficiency check details, see: [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md).

#### 3.1.4 Parameter Check

When a new operator is submitted, the test points are given and the test results are stated.

|                   Test Point                    | Acceptance Standard | Test Result (Error Message) |
| ----------------------------------------------- | --------------------| --------------------------- |
| Whether it conforms to the operator restriction |     Normal error    |                             |
| Whether illegal parameters are passed           |     Normal error    |                             |

### 3.2 Accuracy Test

For the cases used in the New Feature Test section, the features and the number of cases are recorded here. When multiple operations are tested, multiple tables are needed to include details of these operations.

Operation:

|Test Point           | Description                      | Quantity |  Comment |
|----------           |----------------------------------|----------|  --------|
|Data type test       |half/float/int8                   |          |          |
|Mult-tensor test     |Supports 1-8 dims                 |          |          |
|Layout test          |Supports NCHW/NHWC                |          |          |
|Zero element test    |Whether to support this test      |          |          |
|Stability test       |--gtest_repeat=NUM<br>--thread=NUM|          |          |
|Mult-platform test   |MLU370/MLU590                     |          |          |
|Nan/INF test         |Whether to support this test      |          |          |
|Memory leak check    |Test result                       |          |          |
|Code coverage check  |Test result                       |          |          |

### 3.3 Performance Test

See [MLU-OPS Performance Acceptance Standard](https://github.com/Cambricon/mlu-ops/blob/master/docs/MLU-OPS-Performance-Acceptance-Standard.md) for details.

Platform：MLU370

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|-----|----|----|----|-----|
|op_name|    |    |     |    |    |    |     |
|op_name|    |    |     |    |    |    |     |

Platform：MLU590

|Operation|Mlu_hardware_time(us)|Mlu_interface_time(us)|Mlu_io_efficiency|Mlu_compute_efficiency|Mlu_workwpace_size(Bytes)|Data_type|Shape|
|-------|----|----|----|----|----|----|-----|
|op_name|    |    |    |    |    |    |     |
|op_name|    |    |    |    |    |    |     |

### 3.4 Summary Analysis

Please give a brief overview here, if you want to note and summarize the content.

### CODE REVIEW

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 1]
```diff
@@ -0,0 +1,469 @@
...
+### 1.2 算子功能和应用场景描述
+
+- #### 先验知识
+
+一个点云数据集在三维空间中涉及到数据信息包括: 点云自身的位置信息、每个位置对应的特征信息、三维空间大小、体素大小、体数坐标信息，
+其中根据三维空间大小和每个体素大小，可以计算出每个体素在三维空间的坐标，再根据点云的位置信息可以得到每个体素中包含了多个点云数据。
+
+1）voxel_size:[voxel_x, voxel_y, voxel_z]，表示一个体素在x、y、z方向上的长宽高，即体素大小；
+
+2) coors_range:[coors_x_min, coors_y_min, coors_z_min, coors_x_max, coors_y_max, coors_z_max]，给定了体素所在三维空间的边界范围，即三维空间大小；
```
Comment (mahxn0) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:63]: 2没缩进，这里格式预览不对

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 2]
```diff
@@ -0,0 +1,469 @@
...
+\begin{aligned}
+grid\_x = round((coors\_x\_max - coors\_x\_min) / voxel\_x) \\
+grid\_y = round((coors\_y\_max - coors\_y\_min) / voxel\_y) \\
+grid\_z = round((coors\_z\_max - coors\_z\_min) / voxel\_z)
+\end{aligned}
+```
+
+3）`feats`: [N, C], 表示有 N 个点云数据，每个点云有 C 个特征；
+
+4）point_coors: [N, 3], 表示 N 个点云数据对应在三维空间中坐标信息；
```
Comment (mahxn0) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:79]: 统一格式

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 3]
```diff
@@ -0,0 +1,469 @@
...
+\begin{aligned}
+c\_x = floorf((points[idx][0] - coors\_x\_min) / voxel\_x) \\
+c\_y = floorf((points[idx][1] - coors\_y\_min) / voxel\_y) \\
+c\_z = floorf((points[idx][2] - coors\_z\_min) / voxel\_z)
+\end{aligned}
+```
+
+这样每个点对应的体素位置信息(c_x, c_y, c_z)，按照 feats 中点的顺序排序就构成了`dynamic_scatter_forward`算子的输入信息`coors`[N, 3]
+
+说明：由以上推导得到的点在三维空间中体素位置信息`coors`，可以看出一个体素位置(c_x, c_y, c_z)，代表的是空间中一个小长方体，那么在点云数据比较多的情况下，一个体素位置必然包含多个点。
```
Comment (mahxn0) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:93]:
```suggestion
一个体素位置可能包含多个点。
```

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 4]
```diff
@@ -0,0 +1,469 @@
...
+    print(coors_out_mean)
+    feats_out_max, coors_out_max = dsmax(feats, coors)
+    print(feats_out_max)
+    print(coors_out_max)
+
+if __name__ == '__main__':
+    test_dynamic_scatter()
+```
+
+备注：坐标不支持nan/inf。
```
Comment (mahxn0) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:125]: 明确一下　具体哪个输入的哪个维度不支持nan/inf

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+| shape 限制       | voxel_feats、feats的第二维度相等，coors、voxel_coors的第二维度相等                                |
+
+### 1.5 验收标准
+
+#### 1.5.1 精度验收标准
+
+按照[MLU-OPS 算子精度验收标准](../../../MLU-OPS-Accuracy-Acceptance-Standard.md)的要求明确本算子的精度标准。
+本算子为复合类算子：
+- 算子精度验收标准：diff1、diff2；
+- 算子精度阈值描述：diff1 <= 3e-3 && diff2 <=3e-3；
```
Comment (mahxn0) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:170]: 明确每个输出的精度标准，voxel_num,voxel_points_count,point2voxel_map应该是diff3=0?

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 1 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:63]: 已修改

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 2 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:79]: 已修改

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 3 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:93]: 已修改

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 4 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:125]: 已修改

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 5 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:170]: 已修改

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 6]
```diff
@@ -0,0 +1,469 @@
+# RotatedFeatureAlignForward 算子开发设计方案
```
Comment (defei-coder) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:1]: 算子名

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 7]
```diff
@@ -0,0 +1,469 @@
...
+### 1.1 算子需求分析
+
+该需求分析为框架原生算子实现功能的需求分析，对于框架原生支持但 MLU-OPS 当前版本不支持的功能，需要在`1.4算子限制` 章节中显式注明。未明确注明不支持的功能，默认 MLU-OPS 全部支持。
+
+| 算子功能简介           | 将具有相同体素坐标的多个点云数据，在特征维度上利用 `mean` 或 `max` 方法去重为一个点 |
+| ---------------------- | ------------------------------------------------------------ |
+| 需求来源               | mmcv                                                         |
+| 应用网络               | mvxnet                                                       |
+| 输入数据类型           | feats: float32 <br> coors: int32 <br> reduce_type: 枚举类型  |
+| 输入标量参数           | reduce_type: bool                                            |
```
Comment (defei-coder) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:43]: reduce_type 是 枚举类型还是bool？

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 8]
```diff
@@ -0,0 +1,469 @@
...
+| 输入Shape              | feats: [num_points, num_feats] <br> coors: [num_points, 3]   |
+| 输入Layout             | ARRAY                                                        |
+| 输出数据类型           | voxel_feats: float32 <br> voxel_coors: int32 <br> point2voxel_map: int32 <br> voxel_points_count: int32 <br> voxel_num: int32|
+| 输出shape              | voxel_feats: [num_voxels, num_feats] <br> voxel_coors: [num_voxels, 3] <br> point2voxel_map: [num_points] <br> voxel_points_count: [num_voxels] <br> voxel_num: [1]  |
+| 输出Layout             |  ARRAY                                                        |
+| 是否需要支持原位       | 否                                                           |
+| 是否需要支持stride机制 | 否                                                           |
+| 是否需要支持广播       | 否                                                           |
+| 0元素检查是否直接返回  | 输入的维度 num_points 和输出的维度 num_voxels 为 0 时，返回 MLUOP_STATUS_SUCCESS |
```
Comment (defei-coder) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:52]: num_feats 呢？

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 9]
```diff
@@ -0,0 +1,469 @@
...
+
+### 1.2 算子功能和应用场景描述
+
+- #### 先验知识
+
+一个点云数据集在三维空间中涉及到数据信息包括: 点云自身的位置信息、每个位置对应的特征信息、三维空间大小、体素大小、体数坐标信息，
```
Comment (defei-coder) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:58]: 体数坐标 -》体素坐标

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md [hunk 10]
```diff
@@ -0,0 +1,469 @@
...
+```
+2、kernelDynamicScatterForward
+
+```c++
+KERNEL_CHECK((kernelDynamicScatterForward(handle, reduce_type, feats,
+                                          point2voxel_map, num_points,
+                                          num_feats, voxel_points_count，
+                                          voxel_feats)));
+```
+首先计算 nram 上最多可以处理的 points 数量 nram_points_num 。如果 nram_points_num >= 1, 则说明至少可以一次 load 一个整 c 的数据，这样我们按照方式一计算，否则说明 c 很大，无法一次 load, 采用方式二计算。这里的 c = num_feats。
```
Comment (Wickyzheng) [inline docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:257]: 如果C比较小，算子的io量是不是无法打满？也会导致计算量打不满，会有性能问题吧，  可以找框架的同事要下网络规模，在性能测试部分把网络规模补充下

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 6 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:1]: 已修改

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 10 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:263]: 方案二就是当c很小的时候，一次性load nram可以容纳的 c

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 7 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:43]: 是自定义的枚举，文档下面增加了枚举含义

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 9 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:58]: 已修改

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 8 above)
Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:52]: 已修改

File: docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md (hunk 7 above)
Comment (defei-coder) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:43]: reduce_type: bool //这个是？

Comment (tudejiang79) [reply docs/bangc-docs/design_docs/dynamic_scatter_forward/dynamic_scatter_forward.md:43]: 这个我修改下，因为这个值竟品只是实现了两个计算方式，一开始想的是用bool，后来考虑到如果增加了新的计算方式，修改麻烦，就用枚举，43行这个是遗漏修改

**Pull request closed without merging**